src/util/string.h

```diff
@@ -17,6 +17,28 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <int num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { CheckNumFormatSpecifiers(fmt); }
+    consteval static void CheckNumFormatSpecifiers(std::string_view str)
+    {
+        int count{0};
+        bool begin_f{false};
+        for (char c : str) {
+            bool perc{c == '%'};
+            if (begin_f) {
+                count += !perc;
+                begin_f = false;
+                continue;
+            }
+            if (perc) begin_f = true;
+        }
+        if (num_params != count) throw "Number of format specifiers and arguments must match";
```

Comment:
We should be able to simplify this.

-----

Suggestion 1: we swap the state machine when encountering a `%`, otherwise inc the param count and step out of the format:
```C++
constexpr static void CheckNumFormatSpecifiers(std::string_view str)
{
    int count{0};
    bool in_format{false};
    for (char c : str) {
        if (c == '%') in_format = !in_format;
        else if (in_format) {
            ++count;
            in_format = false;
        }
    }
    if (num_params != count) throw std::runtime_error("Format specifier count must match the argument count!");
 }
```

----

Suggestion 2: count the excesses and throw when there is any:
```C++
constexpr static void CheckNumFormatSpecifiers(std::string_view str)
    int excess{num_params};
    bool in_format{false};
    for (char c : str) {
        if (c == '%') in_format = !in_format;
        else if (in_format) {
            --excess;
            in_format = false;
        }
    }
    if (excess) throw std::runtime_error("Format specifier count must match the argument count!");
 }
```

Reply:
By counting the excess we could give more fine-grained errors, e.g:
```C++
constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
{
    int excess{num_params};
    bool in_format{false};
    for (char c: str) {
        if (c == '%') in_format = !in_format;
        else if (in_format) {
            --excess;
            in_format = false;
        }
    }
    if (in_format) throw "Format specifier incorrectly terminated by end of string!";
    if (excess > 0) throw "Too few format specifiers!";
    if (excess < 0) throw "Too many format specifiers!";
}
```

and tests like:
```C++
auto check_too_many_specifiers = [](ErrType str) { return str == "Too many format specifiers!"s; };
BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%_"), ErrType, check_too_many_specifiers);

auto check_too_few_specifiers = [](ErrType str) { return str == "Too few format specifiers!"s; };
BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), ErrType, check_too_few_specifiers);
BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%_"), ErrType, check_too_few_specifiers);
```

Reply:
> important thing is the error, not the exact wording of the error message

You have separate error for trailing `%`, why not have dedicated errors for when the format specifiers are fewer/more than the args?

Reply:
I saw your previous response, I was quoting from it in mine. Anyway, you can resolve this comment.